### PR TITLE
Refine dygraph DataLoader implementation

### PR DESCRIPTION
**Pull request by chenwhql**
Now the dygraph DataLoader will decrease training speed, because of some implementation issues.
This PR refine the dygraph DataLoader implementation to solve following problem.

image

### Optimization effect (single card)

**transformer（s）**
batch_size = 256, epoch = 20

| original DataLoader | New DataLoader | Time decrease | 
|---|---|---|
| 1418.013 | 1320.847 | -6.9% |

**resnet（s）**
batch_size = 256, epoch = 120

| original DataLoader | New DataLoader | Time decrease | 
|---|---|---|
| 12221.662 | 11468.035 | -6.2% |

### CODE REVIEW

File: paddle/fluid/pybind/reader_py.cc [hunk 1]
```diff
@@ -207,6 +209,30 @@ void BindReader(py::module *module) {
            py::call_guard<py::gil_scoped_release>())
       .def("read_next_list", &MultiDeviceFeedReader::ReadNextList,
            py::call_guard<py::gil_scoped_release>())
+      .def("read_next_var_list",
+           [](MultiDeviceFeedReader &self) {
+             auto result_list = self.ReadNextList();
+             std::vector<framework::LoDTensor> tensor_list = result_list[0];
```
Comment (sneaxiy) [inline paddle/fluid/pybind/reader_py.cc]: `auto &tensor_list = result_list[0]`? Do not deep copy.

File: paddle/fluid/pybind/reader_py.cc [hunk 2]
```diff
@@ -207,6 +209,30 @@ void BindReader(py::module *module) {
...
+             std::vector<std::shared_ptr<imperative::VarBase>> var_list;
```
Comment (sneaxiy) [inline paddle/fluid/pybind/reader_py.cc:216]: `var_list.reserve(tensor_list.size());`

File: paddle/fluid/pybind/reader_py.cc [hunk 3]
```diff
@@ -207,6 +209,30 @@ void BindReader(py::module *module) {
...
+               std::string act_name =
+                   imperative::GetCurrentTracer()->GenerateUniqueName(
+                       "generated_var");
+               auto new_var = std::make_shared<imperative::VarBase>(act_name);
+               new_var->SetPersistable(false);
+               new_var->SetType(framework::proto::VarType::LOD_TENSOR);
+               new_var->SetDataType(lod_tensor.type());
+               auto *tensor =
+                   new_var->MutableVar()->GetMutable<framework::LoDTensor>();
+               tensor->ShareDataWith(lod_tensor);
```
Comment (sneaxiy) [inline paddle/fluid/pybind/reader_py.cc]: `*tensor = std::move(lod_tensor);`

File: paddle/fluid/pybind/reader_py.cc [hunk 4]
```diff
@@ -207,6 +209,30 @@ void BindReader(py::module *module) {
...
+               return new_var;
+             };
+             for (auto tensor : tensor_list) {
```
Comment (sneaxiy) [inline paddle/fluid/pybind/reader_py.cc]: `for (auto &tensor : tensor_list)`

File: paddle/fluid/pybind/reader_py.cc (hunk 1 above)
Comment (chenwhql) [reply paddle/fluid/pybind/reader_py.cc]: done

File: paddle/fluid/pybind/reader_py.cc (hunk 2 above)
Comment (chenwhql) [reply paddle/fluid/pybind/reader_py.cc:216]: done

File: paddle/fluid/pybind/reader_py.cc (hunk 3 above)
Comment (chenwhql) [reply paddle/fluid/pybind/reader_py.cc]: done

File: paddle/fluid/pybind/reader_py.cc (hunk 4 above)
Comment (chenwhql) [reply paddle/fluid/pybind/reader_py.cc]: done

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Pull request merged**